File: CSharp_PlayGround/Challenges/LeetCode/RemoveElementNo27.cs
using System;
using System.Linq;


namespace ConsoleApp_PlayGround.Challenges.LeetCode
{
    //URL   
    class RemoveElementClass
    {

        //Runtime 0ms Beats 0%
        //Memory 0MB Beats 0%
        public static int RemoveElement(int[] nums, int val)
        {
            //wrong answer because for just the arrange of elements
            if (nums.Length == 0) return 0;
            //# Select will just change each element to new form
            //var numsFiltered = nums.Select(e => e != val).ToArray();
            var numsFiltered = nums.Where(e => e != val).ToArray();
            numsFiltered.CopyTo(nums, 0);
            return nums.Length - numsFiltered.Length;
        }


        //Runtime 0ms Beats 0%
        //Memory 0 MB Beats 0%
        public static int RemoveElement_Final(int[] nums, int val)
        {
            if (nums.Length == 0) return 0;
            int index = 0;
            for (int i = 0; i < nums.Length; i++)
            {
                if (nums[i] != val)
                {
                    nums[index] = nums[i];
                    ++index;
                }
            }
            return index;
        }

        public static void Implement()
        {          
            int[] nums = [3,2,2,3];
            Console.WriteLine(RemoveElement_Final(nums,3));

            nums = [5,9,5,12,3,5];
            Console.WriteLine(RemoveElement_Final(nums, 5));
        }

    }
}

File: JavaScript_PlayGround/challenges/leetcode/remove-element-No27.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
    if(nums.length === 0) return 0;
    //#Array.map() like c# '[].Select()' will change/transform each element
    //let numsFiltered = nums.map(e => e !== val);
    let numsFiltered = nums.filter(e => e !== val);
    nums.splice(0,numsFiltered.length,...numsFiltered);
    return nums.length  - numsFiltered.length;
};

var removeElement_Final = function(nums, val) {
    if(nums.length === 0) return 0;
    let index= 0;
    for(let i = 0; i < nums.length; i++){
       if(nums[i]!= val){
         nums[index] = nums[i];
         ++index;
       } 
    }
    return index;    
};


let testCase= [5,9,5,12,3,5];
let val= 5;
console.time();
 console.log({value:testCase, result:removeElement(testCase,val)});
console.timeEnd();

testCase = [5,9,5,12,3,5];
console.time();
console.log({value:testCase, result:removeElement_Final(testCase,val)});
console.timeEnd();
 //testCase="()";

